src/analyze-dependencies.ts

```diff
@@ -0,0 +1,270 @@
+              // Not a package or can't read package.json, continue
+            }
+            // Only follow symlinks that point to node_modules
+            if (realPath.includes('node_modules')) {
+              // this.log('Following symlink to:', realPath);
+              await this.walkNodeModules(realPath, callbacks, seenPackages);
+            }
+          } catch {
+            // this.log('Error resolving symlink:', fullPath);
+          }
+          continue;
+        }
+
+        if (entry.isDirectory()) {
+          // Check if this is a package directory
+          const pkgJsonPath = path.join(fullPath, 'package.json');
+          try {
+            const pkgJson = JSON.parse(await fs.readFile(pkgJsonPath, 'utf-8'));
+            // Only process each package once
+            if (!seenPackages.has(pkgJson.name)) {
+              seenPackages.add(pkgJson.name);
+              // this.log('Detected package:', pkgJson.name, 'at', fullPath);
+              callbacks.onPackage(pkgJson);
+            } else {
+              // this.log('Already seen package:', pkgJson.name, 'at', fullPath);
+            }
+          } catch {
+            // Not a package or can't read package.json, continue walking
+          }
+
+          // Continue walking if it's not node_modules
+          if (entry.name !== 'node_modules') {
+            await this.walkNodeModules(fullPath, callbacks, seenPackages);
+          }
+        } else {
+          callbacks.onFile(fullPath);
+        }
+      }
+    } catch {
+      // this.log('Error walking directory:', dir);
+    }
+  }
+}
+
+export class RemoteDependencyAnalyzer implements DependencyAnalyzer {
+  async analyzeDependencies(root: string): Promise<DependencyStats> {
+    const pkgJsonPath = path.join(root, 'package.json');
+    const pkgJson = JSON.parse(await fs.readFile(pkgJsonPath, 'utf-8'));
+
+    // Count direct dependencies
+    const directDependencies = Object.keys(pkgJson.dependencies || {}).length;
+    const devDependencies = Object.keys(pkgJson.devDependencies || {}).length;
+
+    // Analyze dependencies from npm registry
+    const cjsDependencies = 0;
+    const esmDependencies = 0;
+    const installSize = 0;
+
+    // TODO: Implement npm registry fetching
+    // For each dependency:
+    // 1. Fetch package metadata from registry
+    // 2. Analyze module type
+    // 3. Fetch tarball and calculate size
+
+    return {
+      totalDependencies: directDependencies + devDependencies,
+      directDependencies,
+      devDependencies,
+      cjsDependencies,
+      esmDependencies,
+      installSize
+    };
+  }
+}
+
+// Keep the existing tarball analysis for backward compatibility
+export async function analyzeDependencies(
+  tarball: ArrayBuffer
+): Promise<DependencyStats> {
+  const {files, rootDir} = await unpack(tarball);
+  const decoder = new TextDecoder();
+
+  // Set global tarball file list for CLI display
+  globalThis.lastTarballFiles = files.map((f) => f.name);
```

Comment:
can this just be part of the result?

we should avoid using global state i think